pkg/planner/core/operator/logicalop/logical_union_all.go

```diff
@@ -191,7 +193,54 @@ func (p *LogicalUnionAll) ExhaustPhysicalPlans(prop *property.PhysicalProperty)
 
 // CanPushToCop inherits BaseLogicalPlan.LogicalPlan.<21st> implementation.
 
-// ExtractFD inherits BaseLogicalPlan.LogicalPlan.<22nd> implementation.
+// ExtractFD implement base.LogicalPlan.<22nd> interface.
+func (p *LogicalUnionAll) ExtractFD() *fd.FDSet {
+	// basically extract the children's fdSet.
+	childFDs := make([]*fd.FDSet, 0, len(p.children))
+	for _, child := range p.children {
+		childFD := child.ExtractFD()
+		childFDs = append(childFDs, childFD)
+	}
+	// check the output columns' not-null property.
+	res := &fd.FDSet{}
+	notNullCols := intset.NewFastIntSet()
+	for _, col := range p.Schema().Columns {
+		notNullCols.Insert(int(col.UniqueID))
+	}
+	for _, childFD := range childFDs {
+		notNullCols.IntersectionWith(childFD.NotNullCols)
+	}
+	res.MakeNotNull(notNullCols)
+	// check the equivalency between children.
+	for i := 0; i < p.Schema().Len(); i++ {
+		for j := i + 1; j < p.Schema().Len(); j++ {
+			// detect the equivalency between the i-th and j-th column.
+			flag := true
+			iID := int(p.schema.Columns[i].UniqueID)
+			jID := int(p.schema.Columns[j].UniqueID)
+			for _, childFD := range childFDs {
+				equivs := childFD.EquivalenceCols()
+				isEquiv := false
+				for _, equiv := range equivs {
+					if equiv.Has(iID) && equiv.Has(jID) {
+						isEquiv = true
+						break
+					}
+				}
```

Comment:
wrap here as an inplace iteration to avoid the allocation caused by `EquivalenveCols`

Reply:
i changed it to func AreColsEquiv which won't alloc any mem